apps/docs/content/guides/auth/third-party/firebase-auth.mdx

```diff
@@ -0,0 +1,254 @@
+
+</Admonition>
+
+Firebase Auth uses a single set of JWT signing keys for all projects. This means that JWTs issued from an unrelated Firebase project to yours could access data in your Supabase project. To guard against this, creating and maintaining the following RLS policies for **all of your tables in the `public` schema** is very important. You should also attach this policy to [Storage]() buckets or [Realtime]() channels.
+
+To achieve this we recommend using a [restrictive Postgres Row-Level Security policy](https://www.postgresql.org/docs/current/sql-createpolicy.html).
+
+Restrictive RLS policies differ from regular (or permissive) policies in that they use the `as restrictive` clause when being defined. They do not grant permissions, but rather restrict any existing or future permissions. They're great for cases like this where the technical limitations of Firebase Auth remain separate from your app's logic.
+
+<Admonition type="caution">
+
+Postgres has two types of policies: permissive and restrictive. This example uses restrictive policies so make sure you don't omit the `as restrictive` clause.
+
+</Admonition>
+
+This is an example of such an RLS policy that will restrict access to only your project's (denoted with `<firebase-project-id>`) users, and not any other Firebase project.
+
+```sql
+create policy "Restrict access to Supabase Auth and Firebase Auth for project ID <firebase-project-id>"
+  on table_name
+  as restrictive
+  to authenticated
+  using (
+    (auth.jwt()->>'iss' = 'https://<project-ref>.supabase.co/auth/v1')
+    or
+    (
+        auth.jwt()->>'iss' = 'https://securetoken.google.com/<firebase-project-id>'
+        and
+        auth.jwt()->>'aud' = '<firebase-project-id>'
+     )
+  );
+```
+
+If you have a lot of tables in your app, or need to manage complex RLS policies for [Storage]() or [Realtime]() it can be useful to define a [stable Postgres function](https://www.postgresql.org/docs/current/xfunc-volatility.html) that performs the check to cut down on duplicate code. For example:
+
+```sql
+create function public.is_supabase_or_firebase_project_jwt()
+  returns bool
+  language sql
+  stable
+  returns null on null input
+  return (
+    (auth.jwt()->>'iss' = 'https://<project-ref>.supabase.co/auth/v1')
+    or
+    (
+        auth.jwt()->>'iss' = concat('https://securetoken.google.com/<firebase-project-id>')
+        and
+        auth.jwt()->>'aud' = '<firebase-project-id>'
+     )
+  );
+```
+
+Please make sure you substitute `<project-ref>` with your Supabase project's ID and the `<firebase-project-id>` to your Firebase Project ID. Then the restrictive policies on all your tables, buckets and channels can be simplified to be:
+
+```sql
+create policy "Restrict access to correct Supabase and Firebase projects"
+  on table_name
+  as restrictive
+  to authenticated
+  using (public.is_supabase_or_firebase_project_jwt() is true);
```

Comment:
should the function call be prefixed with select for performance? https://supabase.com/docs/guides/database/database-advisors?queryGroups=lint&lint=0003_auth_rls_initplan

Reply:
Huh! Neat trick! Didn't know about this. It should, yes!